File: 020.valid_parentheses/valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> pStack;
        for (char ch: s) {
            switch (ch) {
                case '(':
                case '[':
                case '{':
                    pStack.push(ch);
                    break;
                case ')':
                    if (pStack.empty() || pStack.top() != '(')
                        return false;
                    else
                        pStack.pop();
                    break;
                case ']':
                    if (pStack.empty() || pStack.top() != '[')
                        return false;
                    else
                        pStack.pop();
                    break;
                case '}':
                    if (pStack.empty() || pStack.top() != '{')
                        return false;
                    else
                        pStack.pop();
                    break;
            }
        }
        return pStack.empty();
    }
};

File: 020.valid_parentheses/valid_parentheses.py
class Solution(object):
    def isValid(self, s):
        stack = []
        dict = {"(":")", "[":"]", "{":"}"}
        for i in range(len(s)):
            if s[i] in dict:
                stack.append(s[i])
            else:
                if len(stack) > 0 and s[i] == dict[stack[-1]]:
                    stack.pop()
                else:
                    return False
        return len(stack) == 0
